src/DateScalar.php

```diff
@@ -0,0 +1,138 @@
+<?php declare(strict_types=1);
+
+namespace MLL\GraphQLScalars;
+
+use function assert;
+use DateTimeImmutable;
+use DateTimeInterface;
+use Exception;
+use GraphQL\Error\Error;
+use GraphQL\Error\InvariantViolation;
+use GraphQL\Language\AST\StringValueNode;
+use GraphQL\Type\Definition\ScalarType;
+use GraphQL\Utils\Utils;
+use function is_string;
+use function Safe\preg_match;
+use function Safe\sprintf;
+use function Safe\substr;
+use function strlen;
+use function strpos;
+
+abstract class DateScalar extends ScalarType
+{
+    public function serialize($value): string
+    {
+        if (! $value instanceof DateTimeInterface) {
+            $value = $this->tryParsingDate($value, InvariantViolation::class);
+        }
+
+        return $this->format($value);
+    }
+
+    public function parseValue($value): DateTimeInterface
+    {
+        return $this->tryParsingDate($value, Error::class);
+    }
+
+    public function parseLiteral($valueNode, ?array $variables = null): DateTimeInterface
+    {
+        if (! $valueNode instanceof StringValueNode) {
+            throw new Error(
+                "Query error: Can only parse strings, got {$valueNode->kind}",
+                $valueNode
+            );
+        }
+
+        return $this->tryParsingDate($valueNode->value, Error::class);
+    }
+
+    /**
+     * @param mixed                   $value Any value that might be a Date
+     * @param class-string<Exception> $exceptionClass
+     *
+     * @throws \GraphQL\Error\InvariantViolation|\GraphQL\Error\Error
+     */
+    protected function tryParsingDate($value, string $exceptionClass): DateTimeInterface
+    {
+        if (is_string($value)) {
+            if ('' === $value) {
+                throw new $exceptionClass(
+                    Utils::printSafeJson('Cannot parse empty string')
+                );
+            }
+
+            if (! $this->validateDatetime($value)) {
+                throw new $exceptionClass(
```

Comment:
Is this safe for every `Exception`? Perhaps we should narrow the signature to `Error|InvariantViolation` to ensure a compatible constructor is available.